### PR TITLE
Restore `builtins.pathExists` behavior on broken symlinks

### PR DESCRIPTION
**Pull request by alois31**
Commit 83c067c0fa0cc5a2dca440e5c986afe40b163802 changed `builtins.pathExists` to resolve symlinks before checking for existence. Consequently, if the path refers to a symlink itself, existence of the target of the symlink (instead of the symlink itself) was checked. Restore the previous behavior by skipping symlink resolution in the last component.

# Motivation
Restore backwards compatibility.

# Context
Fixes: https://github.com/NixOS/nix/issues/9901

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Comment by alois31** [comment]
Note that the offending change inadvertedly implemented https://github.com/NixOS/nix/issues/9585, which is effectively reverted here too for more faithful compatibility. If the new behavior is deemed better in this specific case, I could easily adapt the PR.

**Comment by Ericson2314** [comment]
Per https://github.com/NixOS/nix/issues/9585#issuecomment-1938673243 actually 2.18 is wrong, and the newer behavior that makes more sense is also the original older behavior.

**Comment by alois31** [comment]
The obviously correct behavior for the directory thing is now implemented, along with a test to prevent re-regression.

**Comment by Ericson2314** [comment]
@alois31 hmm so I am wondering whether the trailing slash should make a difference, as I believe is how its implemented.

One way to rationalize the trailing `.` behavior is, until the `.` is removed, the symlink *is* an ancestor component.

**Comment by alois31** [comment]
Yeah, that would be my reasoning as well. The comment "SourcePath doesn't know about trailing slash." applies here too.

**Comment by Ericson2314** [comment]
Ah I see. I guess I am contemplating that `coerceToPath` should instead optionally resolve symlinks.

**Comment by alois31** [comment]
Interesting idea. Investigating this exposed *another* regression: `coerceToPath` in addition to stripping the slash also resolves `..` improperly (without taking into account symlinks).

**Comment by Ericson2314** [comment]
@alois31 my understanding is that in fact for many years `..` was resolved incorrectly, but yeah we already started fixing that bug (some nastiness in my personal NixOS setup was relying on it).

**Comment by Ericson2314** [comment]
Checkout https://github.com/NixOS/nix/pull/9881. You can use the callback and a `SourceAccessor` to resolve symlinks properly.

**Comment by thufschmitt** [comment]
Discussed during the Nix maintainers meeting on 2024-02-12.
@ericson2314 will help move this forward (well, he already started).

<details>

- Looks good, just needs someone to check the code.
- Question related to https://github.com/NixOS/nix/issues/9585
    - Should we keep the historical behavior (as reported by the issue) or the 2.20 one?
    - Actually, #9585 is a regression on 2.18, prior versions did return `true` on the example.
    - Conclusion: The current behavior is the right one, keep it
- Assigned to @ericson2314

https://github.com/NixOS/nix/pull/9985#issuecomment-1938674380

https://github.com/NixOS/nix/issues/9585#issuecomment-1938673243
</details>

**Comment by alois31** [comment]
> Check out #9881. You can use the callback and a `SourceAccessor` to resolve symlinks properly.

Hm, `CanonPath` plus `SourceAccessor` begins to smell an awful lot like `SourcePath`. I now wonder if `CanonPath` should just be merged into `SourcePath`…

**Comment by Ericson2314** [comment]
@alois31 `CanonPath` should say separate, but yes this way of making a `CanonPath` could also just be a way of making a `SourcePath` (instead of just temporary using the accessor, it would hold onto it.)

**Comment by Ericson2314** [comment]
Yes, and yes.

File: src/libutil/source-path.hh
Comment (Ericson2314) [inline src/libutil/source-path.hh:14]:
```suggestion
/**
 * Note there is a decent chance this type soon goes away because the problem is solved another way.
 */
enum class SymlinkResolution {
```

File: src/libutil/source-path.hh
Comment (Ericson2314) [inline src/libutil/source-path.hh:123]:
```suggestion
     * given resolution mode.
     *
     *  @param mode might only be a temporary solution for this. 
     * See the discussion in https://github.com/NixOS/nix/pull/9985.
```

**Review by Ericson2314 — APPROVED** [review]
The rest of the Nix team pointed out we want a bug fix that we can backport, so I am approving this as-is after all. Hopefully @alois31 you are still interested working on the other solution in a follow-up :).

**Comment by infinisil** [comment]
@Ericson2314 (regression started in 2.20, so 2.19 backport not necessary)

File: src/libexpr/primops.cc
Comment (edolstra) [inline src/libexpr/primops.cc:118]: BTW, instead of an `std::optional`, wouldn't it be cleaner to have a `SymlinkResolution::None` alternative?

Comment (alois31) [reply src/libexpr/primops.cc:118]: Let's see how much of that footgun remains at all after working on the nicer fix.

**Comment by Ericson2314** [comment]
@alois31 are you still interested in working on the follow-up cleanup?

**Comment by alois31** [comment]
I fear I will not be able to work on this, sorry. The task seems to be much bigger than initially anticipated and I'm not up to it.

**Pull request merged**
